### PR TITLE
Nifi-3855 Add attachment support to PutSlack

### PR DESCRIPTION
**Pull request by tequalsme**
Support including attachments array in message payload, as outlined at https://api.slack.com/docs/message-attachments

### CODE REVIEW

File: nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PutSlack.java [hunk 1]
```diff
@@ -222,7 +273,7 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
                 session.transfer(flowFile, REL_FAILURE);
                 context.yield();
             }
-        } catch (IOException e) {
+        } catch (IOException | JsonParsingException e) {
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PutSlack.java]: This seems like a different error. Even if the behavior (penalize and route to failure) is the same, we probably should have two separate logging messages, to indicate the difference between a connection error vs a "bad data" error.

Comment (tequalsme) [inline nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PutSlack.java]: @mattyb149 Good call, I have updated the code.

File: nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/test/java/org/apache/nifi/processors/slack/PutSlackTest.java [hunk 2]
```diff
@@ -102,6 +106,81 @@ public void testInvalidIconEmoji() {
...
+        props.put("ping", "pong");
+        ff = session.putAllAttributes(ff, props);
+
+        testRunner.setProperty(PutSlack.WEBHOOK_URL, server.getUrl());
+        testRunner.setProperty(PutSlack.WEBHOOK_TEXT, WEBHOOK_TEST_TEXT);
+        PropertyDescriptor dynamicProp = new PropertyDescriptor.Builder()
+                .dynamic(true)
+                .name("foo")
+                .build();
+        testRunner.setProperty(dynamicProp, "{\"foo\": ${foo}, \"ping\":\"${ping}\"}");
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/test/java/org/apache/nifi/processors/slack/PutSlackTest.java]: Are the payloads always key/value pairs? If so I'm thinking the dynamic properties could be the key/value pairs, and the processor could create a JSON payload from them?  If the attachment payload is more flexible, then JSON should be good. Any chance of injection attacks or anything?

Comment (tequalsme) [inline nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/test/java/org/apache/nifi/processors/slack/PutSlackTest.java]: The attachments can get pretty complex. For example, from the [Slack docs](https://api.slack.com/docs/message-attachments):

```
{
            "fallback": "Required plain-text summary of the attachment.",
            "color": "#36a64f",
            "pretext": "Optional text that appears above the attachment block",
            "author_name": "Bobby Tables",
            "author_link": "http://flickr.com/bobby/",
            "author_icon": "http://flickr.com/icons/bobby.jpg",
            "title": "Slack API Documentation",
            "title_link": "https://api.slack.com/",
            "text": "Optional text that appears within the attachment",
            "fields": [
                {
                    "title": "Priority",
                    "value": "High",
                    "short": false
                }
            ],
            "image_url": "http://my-website.com/path/to/image.jpg",
            "thumb_url": "http://example.com/path/to/thumb.png",
            "footer": "Slack API",
            "footer_icon": "https://platform.slack-edge.com/img/default_application_icon.png",
            "ts": 123456789
        }
```

But more importantly, you could specify multiple attachment blocks like the above, displayed in key-sorted order. So I am not sure how that would be handled if the processor is building multiple JSON attachments from individual key/value pairs. As the PR currently stands, the key is used for ordering of attachments - thus the use of SortedSet.

FWIW, my team has been using this code for quite awhile now to create "rich" Slack messages. Slack webhooks will return an error for malformed attachments.

Comment (mattyb149) [inline nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/test/java/org/apache/nifi/processors/slack/PutSlackTest.java]: Sounds good, thanks for the clarification and additional information!

**Comment by mattyb149** [comment]
+1 LGTM, ran build and unit tests, tried against the #test Apache NiFi Slack channel, verified the attachment appeared as specified.  Nice improvement, thanks very much for the contribution! Merging to master

**Pull request closed without merging**